mods/default/furnace.lua

```diff
@@ -109,61 +111,71 @@ local function furnace_node_timer(pos, elapsed)
 	local fuel_totaltime = meta:get_float("fuel_totaltime") or 0
 
 	local inv = meta:get_inventory()
-	local srclist = inv:get_list("src")
-	local fuellist = inv:get_list("fuel")
-
-	--
-	-- Cooking
-	--
-
-	-- Check if we have cookable content
-	local cooked, aftercooked = minetest.get_craft_result({method = "cooking", width = 1, items = srclist})
-	local cookable = true
-
-	if cooked.time == 0 then
-		cookable = false
-	end
-
-	-- Check if we have enough fuel to burn
-	if fuel_time < fuel_totaltime then
-		-- The furnace is currently active and has enough fuel
-		fuel_time = fuel_time + 1
-
-		-- If there is a cookable item then check if it is ready yet
-		if cookable then
-			src_time = src_time + 1
-			if src_time >= cooked.time then
-				-- Place result in dst list if possible
-				if inv:room_for_item("dst", cooked.item) then
-					inv:add_item("dst", cooked.item)
-					inv:set_stack("src", 1, aftercooked.items[1])
-					src_time = 0
+	local srclist, fuellist
+
+	local cookable, cooked
+
+	local update = true
+	while update do
+		update = false
+
+		srclist = inv:get_list("src")
+		fuellist = inv:get_list("fuel")
+
+		--
+		-- Cooking
+		--
+
+		-- Check if we have cookable content
+		local aftercooked
+		cooked, aftercooked = minetest.get_craft_result({method = "cooking", width = 1, items = srclist})
```

Comment:
> There's a minetest.register_on_craft function, maybe a mod can use it to play a sound instead of removing the source item(s) when cooking

That's a good argument with those crafting callbacks, but isn't your first example still  one for moving it out of the loop? You would want one sound per update when they are cooking, but also just one, when catching up, instead of 99 (or more) per furnace at once.

But the rarity thing you mention would be messed up instead. :confused:

I think the core problem here is, that `minetest.register_on_craft` is not very clear on how it stands to bulk-crafts (like the middle mouse button in a conventional crafting grid, or a furnace, autocrafter or maybe one day technic machine that catches up).